exporter/uptraceexporter/config.go

```diff
@@ -0,0 +1,25 @@
+// Copyright 2021 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package uptraceexporter
+
+import "go.opentelemetry.io/collector/config/configmodels"
+
+type Config struct {
+	configmodels.ExporterSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
+
+	// DSN is a data source name for your Uptrace project.
+	// Example: https://<key>@api.uptrace.dev/<project_id>
+	DSN string `mapstructure:"dsn"`
```

Comment:
We use endpoint for all the configs, is there anything special here that you need to name it `dns`?

Reply:
Endpoint kinds implies the format that is used by the other receivers, like `hostname:port`. If this property is changed to be named `endpoint`, it should follow the same convention.

Reply:
@jpkrohling good call, but I would like to understand if this exporter only works with a DNS address or we can have a simple endpoint that can also be a DNS entry if needed.

Reply:
Note that DSN != DNS. A DSN (data source name) is typically a URL-like string, and is used typically for database connection strings.

For this here, the DSN should be broken down into different components, like we have for other exporters. Also, the HTTP Client should also be built like the others, to take advantage of common facilities. Not sure how I could have missed this!

Reply:
>is there anything special here that you need to name it dns?

It is a DSN or a connection string. Is there anything wrong it? Is it banned in Otel collector?

>For this here, the DSN should be broken down into different components

What components? How this makes anyone's life better or simpler?

Uptrace UI provides single string to control access. How breaking that single string into multiple components for Uptrace exporter in Otel collector makes any sense or improves anything?

>We use endpoint for all the configs
> like we have for other exporters

What do you mean by this? They all provide different ways to configure their exporters...

- honeycomb has `api_key`, `dataset`, `api_url`
- jaeger has `url`, `timeout`, `headers`
- newrelic has `spans_url_override`, `timeout`, `apikey`

>Also, the HTTP Client should also be built like the others

Do you care to point me to an example in Honeycomb/Newrelic/Jaeger exporters? Or some documentation?

Reply:
> What components? How this makes anyone's life better or simpler?

Protocol, hostname+port. I assume those are the only things that might change. For users of the OpenTelemetry Collector, having your exporter to follow what other exporters are doing is making their lives easier.

> Uptrace UI provides single string to control access. How breaking that single string into multiple components for Uptrace exporter in Otel collector makes any sense or improves anything?

If your users are used to this concept, then I'm certainly fine in letting you use it.

> What do you mean by this? All exporters provide different configurations keys...

I can only speak for Jaeger, and we have only the `endpoint` to control the destination. Not sure which URL you are talking about.

https://github.com/open-telemetry/opentelemetry-collector/blob/68c5961f7bc2d8d55fd431ab07367fc16f8b26d0/exporter/jaegerexporter/config.go#L24-L31

https://github.com/open-telemetry/opentelemetry-collector/blob/68c5961f7bc2d8d55fd431ab07367fc16f8b26d0/config/configgrpc/configgrpc.go#L67

> Do you care to point me to an example in Honeycomb/Newrelic/Jaeger exporters? Or some documentation?

Again, I can only speak for Jaeger, but here's how Jaeger builds a gRPC client connection:

https://github.com/open-telemetry/opentelemetry-collector/blob/68c5961f7bc2d8d55fd431ab07367fc16f8b26d0/exporter/jaegerexporter/exporter.go#L43-L51

It uses the common gRPC configuration object, so that every new property added to that common config will automatically be available to the Jaeger exporter. Similarly, all the TLS configuration is handled by the common config object, providing a single experience to the users.

In my opinion, it's ultimately your decision how you want your exporter to look like, but it should feel like any other exporter whenever possible.

Reply:
> Not sure which URL you are talking about.

I am talking about [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/jaegerthrifthttpexporter/testdata/config.yaml#L8-L15) and it looks rather different even though  it is also Jaeger exporter...

Reply:
True. To be fair, that's quite old and would have been written following the current standards. Good thing is that this isn't the recommended exporter :-)